filters/auth/grantconfig.go

```diff
@@ -0,0 +1,296 @@
+		return c.initErr
+	}
+
+	if c.ClientID == "" && c.ClientIDFile == "" {
+		c.initErr = ErrMissingClientID
+		return c.initErr
+	}
+
+	if c.ClientSecret == "" && c.ClientSecretFile == "" {
+		c.initErr = ErrMissingClientSecret
+		return c.initErr
+	}
+
+	if c.CallbackPath == "" {
+		c.CallbackPath = defaultCallbackPath
+	}
+
+	if c.TokenCookieName == "" {
+		c.TokenCookieName = defaultTokenCookieName
+	}
+
+	if c.TokeninfoClient == nil {
+		client, err := newAuthClient(
+			c.TokeninfoURL,
+			"granttokeninfo",
+			c.ConnectionTimeout,
+			c.MaxIdleConnectionsPerHost,
+			c.Tracer,
+		)
+		if err != nil {
+			return err
+		}
+		c.TokeninfoClient = client
+	}
+
+	if c.AuthClient == nil {
+		c.AuthClient = net.NewClient(net.Options{})
+	}
+
+	c.flowState = newFlowState(c.Secrets, c.SecretFile)
+
+	if c.ClientIDFile != "" {
+		c.SecretsProvider.Add(c.ClientIDFile)
+	}
+
+	if c.ClientSecretFile != "" {
+		c.SecretsProvider.Add(c.ClientSecretFile)
+	}
+
+	c.initialized = true
+	return nil
+}
+
+func (c *OAuthConfig) NewGrant() (filters.Spec, error) {
+	if err := c.init(); err != nil {
+		return nil, err
+	}
+
+	return &grantSpec{config: *c}, nil
+}
+
+func (c *OAuthConfig) NewGrantCallback() (filters.Spec, error) {
+	if err := c.init(); err != nil {
+		return nil, err
+	}
+
+	return &grantCallbackSpec{config: *c}, nil
+}
+
+func (c *OAuthConfig) NewGrantClaimsQuery() (filters.Spec, error) {
+	if err := c.init(); err != nil {
+		return nil, err
+	}
+
+	return &grantClaimsQuerySpec{
+		oidcSpec: oidcIntrospectionSpec{
+			typ: checkOIDCQueryClaims,
+		},
+	}, nil
+}
+
+func (c *OAuthConfig) NewGrantPreprocessor() (routing.PreProcessor, error) {
+	if err := c.init(); err != nil {
+		return nil, err
+	}
+
+	return grantPrep{config: *c}, nil
+}
+
+func (c *OAuthConfig) GetConfig() *oauth2.Config {
+	return &oauth2.Config{
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  c.AuthURL,
+			TokenURL: c.TokenURL,
+		},
+		ClientID:     c.GetClientID(),
+		ClientSecret: c.GetClientSecret(),
+	}
+}
+
+func (c *OAuthConfig) GetAuthURLParameters(redirectURI string) []oauth2.AuthCodeOption {
+	params := []oauth2.AuthCodeOption{oauth2.SetAuthURLParam("redirect_uri", redirectURI)}
+
+	if c.AuthURLParameters != nil {
+		for k, v := range c.AuthURLParameters {
+			params = append(params, oauth2.SetAuthURLParam(k, v))
+		}
+	}
+
+	return params
+}
+
+func (c *OAuthConfig) GetClientID() string {
+	if c.ClientID != "" {
+		return c.ClientID
+	}
+
+	if id, exists := c.SecretsProvider.GetSecret(c.ClientIDFile); exists {
+		return string(id)
+	}
+
+	return ""
+}
+
+func (c *OAuthConfig) GetClientSecret() string {
+	if c.ClientSecret != "" {
+		return c.ClientSecret
+	}
+
+	if secret, exists := c.SecretsProvider.GetSecret(c.ClientSecretFile); exists {
+		return string(secret)
+	}
+
+	return ""
+}
+
+// RedirectURLs constructs the redirect URI based on the request and the
+// configured CallbackPath.
+func (c OAuthConfig) RedirectURLs(req *http.Request) (redirect, original string) {
```

Comment:
Should be pointer `(c *OAuthConfig)`

Reply:
I think makes sense to check elsewhere (e.g. `cookie`) that pointer to struct is used, see https://dave.cheney.net/2016/03/19/should-methods-be-declared-on-t-or-t

> In short, I think that you should prefer declaring methods on *T unless you have a strong reason to do otherwise.

Reply:
Fixed

Reply:
Basically everywhere it is not a pointer (e.g. filter structs) they will be fully copied on each call as
> Methods in Go are just syntactic sugar for a function which passes the receiver as the first formal parameter.

Reply:
👍  ok I updated `grant`, `grantCallback`, `grantPrep`, `cookie`, and their specs to use receiver pointers in functions instead.